group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/MultiThreadedEventProcessor.java

```diff
@@ -0,0 +1,212 @@
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.coordinator.group.runtime;
+
+import org.apache.kafka.common.utils.LogContext;
+import org.slf4j.Logger;
+
+import java.util.List;
+import java.util.concurrent.RejectedExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+/**
+ * A multithreaded {{@link CoordinatorEvent}} processor which uses a {{@link EventAccumulator}}
+ * which guarantees that events sharing a partition key are not processed concurrently.
+ */
+public class MultiThreadedEventProcessor implements CoordinatorEventProcessor {
+
+    /**
+     * The logger.
+     */
+    private final Logger log;
+
+    /**
+     * The accumulator.
+     */
+    private final EventAccumulator<Integer, CoordinatorEvent> accumulator;
+
+    /**
+     * The processing threads.
+     */
+    private final List<EventProcessorThread> threads;
+
+    /**
+     * The lock for protecting access to the resources.
+     */
+    private final ReentrantLock lock;
+
+    /**
+     * A boolean indicated whether the event processor is shutting down.
+     */
+    private volatile boolean shuttingDown;
+
+    /**
+     * Constructor.
+     *
+     * @param logContext    The log context.
+     * @param threadPrefix  The thread prefix.
+     * @param numThreads    The number of threads.
+     */
+    public MultiThreadedEventProcessor(
+        LogContext logContext,
+        String threadPrefix,
+        int numThreads
+    ) {
+        this.log = logContext.logger(MultiThreadedEventProcessor.class);
+        this.shuttingDown = false;
+        this.lock = new ReentrantLock();
+        this.accumulator = new EventAccumulator<>();
+        this.threads = IntStream.range(0, numThreads).mapToObj(threadId ->
+            new EventProcessorThread(
+                threadPrefix + threadId
+            )
+        ).collect(Collectors.toList());
+        this.threads.forEach(EventProcessorThread::start);
+    }
+
+    /**
+     * The event processor thread. The thread pulls events from the
+     * accumulator and runs them.
+     */
+    class EventProcessorThread extends Thread {
+        private final Logger log;
+
+        EventProcessorThread(
+            String name
+        ) {
+            super(name);
+            log = new LogContext("[" + name + "]: ").logger(EventProcessorThread.class);
+            setDaemon(false);
+        }
+
+        private void handleEvents() {
+            while (!shuttingDown) {
+                CoordinatorEvent event = accumulator.poll();
```

Comment:
do we need the lock in enqueue/shutdown?

> prevent the queue from never emptying

once we set shuttingDown=true in `beginShutdown()`, shouldn't the other threads learn to `drainEvents()` without locks?

Reply:
> I noticed we lock on enqueuing and shutdown, but not when we remove from the accumulator. Is the main reason for the lock to prevent the queue from never emptying when shutting down?

Right. I use the lock to protect `shuttingDown` and `accumulator` together. The `accumulator` is thread safe so we don't need to protect it when accessing it alone.

> once we set shuttingDown=true in beginShutdown(), shouldn't the other threads learn to drainEvents() without locks?

Right. Note that the threads do not acquire the lock.

As I second though, I think that I can avoid the lock in `enqueue` but I need to close the accumulator before setting shuttingDown to true to ensure that new events are rejected before threads start draining the accumulator. Do you agree?

Reply:
Seems reasonable to me.

Reply:
Do we need the lock at all then? I see it still in beginShutdown and close.

Reply:
is it more on logging multiple times? i don't see any harm in calling accumulator.close multiple times. also on MultiThreadedEventProcessor#close()

Reply:
Removed the lock. I have put `synchronized` to `beginShutdown` instead. This is needed because `shuttingDown` is accessed twice for the logging message as you pointed out.

Reply:
can MultiThreadedEventProcessor#close() be called multiple times? we also have logging there as well. if beginShutdown is only called through close() then maybe we can synchronize there

Reply:
makes sense